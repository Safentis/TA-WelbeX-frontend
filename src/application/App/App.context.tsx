import { createContext, useContext } from 'react';

import { Context } from './App.interface';

export const AppContext = createContext<Context>({
  state: {
    filter: {
      column: {
        name: true,
        number: false,
        distance: false,
      },
      operation: '',
      search: '',
    }
  },
  setState: () => {},
});

export const useAppContext = () => useContext(AppContext);
