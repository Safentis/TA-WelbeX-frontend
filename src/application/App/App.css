.application__inner {
    margin: 10rem 15rem;
}

.application__table {
    display: grid;
    grid-template-columns: 1fr;
    grid-template-rows: .2fr 1fr;
    grid-row-gap: 2rem;
}