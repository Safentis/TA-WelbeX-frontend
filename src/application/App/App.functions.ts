import {
  Entry,
  FilterFunction,
  PaginationCalcFunc,
  Params,
} from './App.interface';

/**
 * Function makes filtering by entries
 * @param {Params} params - required, params of filter
 * @param {Entry[]} entries - required, no filtered entries
 * @returns {Entry[]} returns the filtered entries
 */
// prettier-ignore
export const filterFunction = ({ params, entries }: FilterFunction): Entry[] => {
  const { column, operation, search }: Params = params;

  const isColumnChoosed: boolean = column.length > 0;
  const isOperationChoosed: boolean = operation.length > 0;
  const isSearchFill: boolean = search.length > 0;

  //* If user choosed "column" and "operations"
  //* we will be filtering, else nothing
  if (!isColumnChoosed || !isOperationChoosed) {
    return entries;
  }
  //* If search input is fill
  //* we will be returns result
  //* else return prev state
  if (isSearchFill) {
    const result: Entry[] = entries.filter((entry: Entry) => {
      switch (operation) {
        case 'more': //* return all more then search
          return search < entry[column];
        case 'less': //* returns all less then search
          return search > entry[column];
        case 'equal': //* strong compare
          //* If search is NaN we are making compare by
          //* string, else by number   
          return isNaN(+search)
            ? search === entry[column]
            : +search === entry[column];
        case 'contains':
          const str: string = entry[column].toString();
          return str.includes(search);
        default:
          return {};
      }
    });

    return result;
  } else {
    return entries;
  }
};

export const paginationCalcFunc = ({
  currentPage,
  entriesOnPage,
  entries,
}: PaginationCalcFunc): Entry[] => {
  let lastIndex = currentPage * entriesOnPage;
  let firstIndex = lastIndex - entriesOnPage;
  let currentEntries = entries.slice(firstIndex, lastIndex);

  return currentEntries;
};
