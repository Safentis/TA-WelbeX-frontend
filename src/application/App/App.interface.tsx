export interface Props {}

export interface Entry {
  date: string;
  distance: number;
  id: number;
  name: string;
  number: number;
}

export interface State {
  filter: {
    column: {
      [key: string]: boolean;
    },
    operation: string,
    search: string,
  }
}

export interface Context {
  state: State;
  setState: Function;
}