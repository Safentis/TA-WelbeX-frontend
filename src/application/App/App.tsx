import React, { FC, useState } from 'react';

import Sort from '../../layouts/Sort';
import THead from '../../components/THead/THead';
import TBody from '../../components/TBody/TBody';

import './App.css';
import { Entry, Props, State } from './App.interface';
import { AppContext } from './App.context';

/**
 * App is main component that contains others part of application
 * @returns {React.ReactElement} - returns react element
 */
const App: FC<Props> = ({}): React.ReactElement => {
  //* ---------------------------------------------
  //* Main state of application
  const [state, setState] = useState<State>({
    filter: {
      column: {
        name: true,
        number: false,
        distance: false,
      },
      operation: 'equal',
      search: '',
    },
  });

  return (
    <AppContext.Provider value={{state, setState}}>
      <div className="application">
        <div className="application__inner">
          <div className="table application__table">
            <div className="table__functions">
              <Sort />
            </div>
            <table className="table__field">
              <THead />
              <TBody />
            </table>
          </div>
        </div>
      </div>
    </AppContext.Provider>
  );
};

export default App;
