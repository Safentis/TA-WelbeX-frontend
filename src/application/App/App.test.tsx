import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import App from './App';

import { Props } from './App.interface';

const setUp = (props: Props) => {
  return shallow(<App {...props} />);
};

describe('<App />', () => {
  let component: ShallowWrapper<
    any,
    Readonly<{}>,
    React.Component<{}, {}, any>
  >;

  beforeEach(() => {
    component = setUp({});
  });

  it('component does match to the snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  describe('Markup: ', () => {
    it('has 1 table element', () => {
      expect(component.find('table')).toBeTruthy();
    });
    
    it('has 1 Filter element', () => {
      expect(component.find('Filter')).toBeTruthy();
    });

    it('has 1 THead element', () => {
      expect(component.find('THead')).toBeTruthy();
    });

    it('has 1 TBody element', () => {
      expect(component.find('TBody')).toBeTruthy();
    });
  });
});
