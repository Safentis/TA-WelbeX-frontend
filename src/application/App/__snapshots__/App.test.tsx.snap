// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> component does match to the snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "setState": [Function],
      "state": Object {
        "filter": Object {
          "column": Object {
            "distance": false,
            "name": true,
            "number": false,
          },
          "operation": "equal",
          "search": "",
        },
      },
    }
  }
>
  <div
    className="application"
  >
    <div
      className="application__inner"
    >
      <div
        className="table application__table"
      >
        <div
          className="table__functions"
        >
          <Sort />
        </div>
        <table
          className="table__field"
        >
          <THead />
          <TBody />
        </table>
      </div>
    </div>
  </div>
</ContextProvider>
`;
